optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/index/IndexVariableListener.java

```diff
@@ -66,38 +73,56 @@ public void afterEntityRemoved(ScoreDirector<Solution_> scoreDirector, Object en
         // Do nothing
     }
 
-    protected void insert(InnerScoreDirector<Solution_, ?> scoreDirector, Object sourceEntity) {
-        List<Object> listVariable = sourceListVariableDescriptor.getListVariable(sourceEntity);
+    protected void insert(InnerScoreDirector<Solution_, ?> scoreDirector, Object entity) {
+        List<Object> listVariable = sourceVariableDescriptor.getListVariable(entity);
         if (listVariable == null) {
             return;
         }
         int index = 0;
-        for (Object shadowEntity : listVariable) {
+        for (Object value : listVariable) {
             // TODO maybe update inverse relation variable (if exists) to avoid an extra loop
             //  in a dedicated inverse relation listener
-            if (!Objects.equals(indexShadowVariableDescriptor.getValue(shadowEntity), index)) {
-                scoreDirector.beforeVariableChanged(indexShadowVariableDescriptor, shadowEntity);
-                indexShadowVariableDescriptor.setValue(shadowEntity, index);
-                scoreDirector.afterVariableChanged(indexShadowVariableDescriptor, shadowEntity);
+            Object oldIndex = shadowVariableDescriptor.getValue(value);
+            if (oldIndex != null) {
+                throw new IllegalStateException("The entity (" + entity
+                        + ") has a list variable (" + sourceVariableDescriptor.getVariableName()
+                        + ") and one of its values (" + value
+                        + ") which has a shadow variable (" + shadowVariableDescriptor.getVariableName()
+                        + ") has an oldIndex (" + oldIndex + ") which is not null.\n"
+                        + "Verify the consistency of your input problem for that shadow variable.");
```

Comment:
nitpick: we typically start the advise line after the \n with "Maybe", so "Maybe verify ...".
Why? Because the first line is 100% sure: that is the problem, that is the error. The second line, the advise, is never 100% sure. It could be something else too, in theory at least. The "maybe" signals to users that we're not guaranteeing that this advise always works, we just think it works most of the time.

Reply:
Correct but this is currently consistent with rest of the built-in listeners. I've filed https://issues.redhat.com/browse/PLANNER-2619 to review all of them.